worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,1033 @@
+	block::{BlockHash, BlockNumber},
+	traits::{Block as BlockT, SignedBlock as SignedBlockT},
+};
+/// key value of sidechain db of last block
+const LAST_BLOCK_KEY: &[u8] = b"last_sidechainblock";
+/// key value of the stored shards vector
+const STORED_SHARDS_KEY: &[u8] = b"stored_shards";
+
+/// ShardIdentifier type
+type ShardIdentifierFor<B> = <<B as SignedBlockT>::Block as BlockT>::ShardIdentifier;
+/// Helper struct, contains the blocknumber
+/// and blockhash of the last sidechain block
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, Debug, Default)]
+pub struct LastSidechainBlock {
+	/// hash of the last sidechain block
+	pub hash: H256,
+	/// block number of the last sidechain block
+	pub number: BlockNumber,
+}
+
+/// Struct used to insert newly produced sidechainblocks
+/// into the database
+pub struct SidechainStorage<SignedBlock: SignedBlockT> {
+	/// database
+	db: SidechainDB,
+	/// shards in database
+	shards: Vec<ShardIdentifierFor<SignedBlock>>,
+	/// map to last sidechain block of every shard
+	last_blocks: HashMap<ShardIdentifierFor<SignedBlock>, LastSidechainBlock>,
+}
+
+impl<SignedBlock: SignedBlockT> SidechainStorage<SignedBlock> {
+	/// loads the DB from the given paths and stores the listed shard
+	/// and their last blocks in memory for better performance
+	pub fn new(path: PathBuf) -> Result<SidechainStorage<SignedBlock>> {
+		// load db
+		let db = SidechainDB::open_default(path)?;
+		let mut storage = SidechainStorage { db, shards: vec![], last_blocks: HashMap::new() };
+		storage.shards = storage.load_shards_from_db()?;
+		// get last block of each shard
+		for shard in storage.shards.iter() {
+			if let Some(last_block) = storage.load_last_block_from_db(shard)? {
+				storage.last_blocks.insert(*shard, last_block);
+			} else {
+				// an empty shard sidechain storage should not exist. Consider deleting this shard from the shards list.
+				error!("Sidechain storage of shard {:?} is empty", shard);
+			}
+		}
+		Ok(storage)
+	}
+
+	/// gets all shards of currently loaded sidechain db
+	pub fn shards(&self) -> &Vec<ShardIdentifierFor<SignedBlock>> {
+		&self.shards
+	}
+
+	/// gets the last block of the current sidechain DB and the given shard
+	pub fn last_block_of_shard(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+	) -> Option<&LastSidechainBlock> {
+		self.last_blocks.get(shard)
+	}
+
+	/// gets the block hash of the sidechain block of the given shard and block number, if there is such a block
+	pub fn get_block_hash(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+		block_number: BlockNumber,
+	) -> Result<Option<BlockHash>> {
+		self.db.get((*shard, block_number))
+	}
+
+	/// gets the block of the given blockhash, if there is such a block
+	#[allow(unused)]
+	pub fn get_block(&self, block_hash: &BlockHash) -> Result<Option<SignedBlock>> {
+		self.db.get(block_hash)
+	}
+
+	/// update sidechain storage
+	pub fn store_blocks(&mut self, blocks_to_store: Vec<SignedBlock>) -> Result<()> {
```

Comment:
Generally, I find it hard to read code that is many levels of indentations. Loops always imply indentations to simply understanding to the code, I'd generally prefer to have a pattern like:

```
fn store_blocks(blocks: Vec<Block>) {
    for each:
         store_block(block: Block);
}

```
This makes code much easier to read. This is not something you need to fix, but just a general remark.

Reply:
Okay, made the for loop smaller, however left is as a for loop. Seemed nice enough to me. :)